### PR TITLE
Feature findCID endpoint

### PR DESCRIPTION
**Pull request by Pl8tinium**
allow nodes to cache all ipfs related vsc content and make it available via the graphql api

useful for e.g. block explorers

### CODE REVIEW

**Comment by Pl8tinium** [comment]
ive prepended an ipfs pin to every ipfs cat/ dag.get call, so nodes that are responsible for distributing vsc state (vsc tx, vsc blocks, contract code etc) have everything cached locally
(to be enabled per config)

File: src/modules/api/graphql/resolvers.ts [hunk 1]
```diff
@@ -104,5 +143,32 @@ export const Resolvers = {
...
+  // finds and tags vsc-tx/ vsc-blocks via ipfs CID's, unidentifiable CID's are tagged with the type 'null'
+  findCID: async (_, args) => {
+    if (appContainer.self.config.get('ipfs.pinEverything')) {
+      const ipfsHash = args.id;
+  
+      if (!ipfsHash) {
+        return null;
+      }
+  
+      if (ipfsHash.startsWith('Qm') || ipfsHash.startsWith('bafy')) {
```
Comment (techcoderx) [inline src/modules/api/graphql/resolvers.ts:157]: Is there anything else (other than contract manifest/code) that uses DAG-PD (0x70)? Everything else I've seen are using DAG-CBOR (0x71) that only begins with `bafyrei`.

Perhaps doing a `CID.parse()` in a try/catch then checking if code is 0x71 should be the first things to validate.

File: src/modules/api/graphql/resolvers.ts [hunk 2]
```diff
@@ -11,6 +10,46 @@ export const DebugResolvers = {
...
+
+    const buffer = Buffer.concat(chunks);
+    let dataRaw = buffer.toString(); 
+
+    content = tryJSONParse(dataRaw);
+  } else {
+    content = content.value
+  }
+
+  if (typeof content === 'object') {
```
Comment (techcoderx) [inline src/modules/api/graphql/resolvers.ts:42]: Might need to detect any infinite recursion as contents of a CBOR JSON could end up reference its own CID

**Review by techcoderx — COMMENTED** [review]
Haven't test it myself yet, other than some validation checks it looks good to me so far.

Besides `vsc-tx` and `vsc-block` it probably needs to be able to find `multisig_txref` CID as well.

File: src/modules/api/graphql/resolvers.ts (hunk 2 above)
Comment (Pl8tinium) [reply src/modules/api/graphql/resolvers.ts]: great catch! implemented ✌️

File: src/modules/api/graphql/resolvers.ts (hunk 1 above)
Comment (Pl8tinium) [reply src/modules/api/graphql/resolvers.ts]: not that im aware of atm

changed the code like you suggested, now it's way more readable!

**Review by techcoderx — APPROVED** [review]

**Comment by vaultec81** [comment]
I will review this sometime after HiveFest

File: src/services/chainBridge.ts [hunk 3]
```diff
@@ -1053,28 +1058,37 @@ export class ChainBridge {
...
+          //   hive_account: block.tx.posting
+          // }
+        })
+
+        for (let tx of blockContent.txs) {
+
+          // pla: pin vsc tx on enabled nodes, note: every dag.get/ cat is a pin via an overridden base function
+          if (this.self.config.get('ipfs.pinEverything')) {
+            const signedTransaction: DagJWS = (await this.self.ipfs.dag.get(CID.parse(tx.id))).value;
+            (await this.self.ipfs.dag.get((signedTransaction as any).link as CID)).value
```
Comment (vaultec81) [inline src/services/chainBridge.ts:1088]: Same here

File: src/services/chainBridge.ts [hunk 4]
```diff
@@ -461,6 +461,11 @@ export class ChainBridge {
       } catch (err) {
         this.self.logger.error('not able to inject contract into the local database', tx.transaction_id)
       }
+
+      // pla: pin contract code on enabled nodes, note: every dag.get/ cat is a pin via an overridden base function
+      if (this.self.config.get('ipfs.pinEverything')) {
+        for await (const chunk of this.self.ipfs.cat(json.code)) {}
```
[the remark this answers is not in the document]
Comment (Pl8tinium) [reply src/services/chainBridge.ts]: ofc

**Pull request merged**